src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs

```diff
@@ -560,23 +597,24 @@ private void GetItemsAndContinue(EvalResultDataItem dataItem, DkmWorkList workLi
             var numRows = rows.Count;
             Debug.Assert(count >= numRows);
             var results = new DkmEvaluationResult[numRows];
-            GetEvaluationResultsAndContinue(rows, results, 0, numRows, workList, inspectionContext, value.StackFrame,
-                () => ContinueWithExceptionHandling(
+            var wl = new WorkList(workList, e => completionRoutine(DkmEvaluationEnumAsyncResult.CreateErrorResult(e)));
+            GetEvaluationResultsAndContinue(rows, results, 0, numRows, wl, inspectionContext, value.StackFrame,
+                () => wl.ContinueWith(
                     () =>
                     {
                         completionRoutine(new DkmEvaluationEnumAsyncResult(results));
                         rows.Free();
-                    },
-                    e => completionRoutine(DkmEvaluationEnumAsyncResult.CreateErrorResult(e))));
+                    }));
+            wl.Execute();
         }
 
-        private void StoreResultAndContinue(DkmEvaluationResult result, ArrayBuilder<EvalResultDataItem> rows, DkmEvaluationResult[] results, int index, int numRows, DkmWorkList workList, DkmInspectionContext inspectionContext, DkmStackWalkFrame stackFrame, CompletionRoutine completionRoutine)
+        private void StoreResultAndContinue(DkmEvaluationResult result, ArrayBuilder<EvalResultDataItem> rows, DkmEvaluationResult[] results, int index, int numRows, WorkList workList, DkmInspectionContext inspectionContext, DkmStackWalkFrame stackFrame, CompletionRoutine completionRoutine)
```

Comment:
completionRoutine is just the cleanup code.  Another way to express this would be to just call it after wl.Execute().  You might have to do some extra work to make sure that exceptions are handled, but it would probably be clearer than threading the cleanup code through all the calls and continuations.